### PR TITLE
Refactor and allow base64-encoded autocrypt key export

### LINKED ISSUE #1730: Add support for base64-encoded autocrypt keys.
## Description

Currently we support import/export of such keys only in binary format, however it could be handy to import/export base64 keys in one call.

Function `rnp_import_keys()` should be updated to allow just base64-encoded data (which is used to transfer autocrypt keys), without OpenPGP headers. Maybe with additional flag (RNP_ALLOW_BASE64?).

`rnp_key_export_autocrypt()` should be updated as well.

### PR DESCRIPTION
**Pull request by ni4**
This PR:
- refactors armoring/dearmoring, adding some C++ wrappers.
- refactors out unused functions write_pgp_key()/write_pgp_keys()
- adds support for export of base64-encoded autocrypt keys, fixes #1730

### CODE REVIEW

File: src/librepgp/stream-armor.cpp [hunk 1]
```diff
@@ -918,36 +930,34 @@ armored_encode3(uint8_t *out, uint8_t *in)
...
-    } catch (const std::exception &e) {
-        RNP_LOG("%s", e.what());
-        return RNP_ERROR_BAD_STATE;
+    bool base64 = param->type == PGP_ARMORED_BASE64;
+    if (!base64) {
+        try {
+            param->crc_ctx.add(buf, len);
+        } catch (const std::exception &e) {
+            RNP_LOG("%s", e.what());
+            return RNP_ERROR_BAD_STATE;
```
Comment (ronaldtse) [inline src/librepgp/stream-armor.cpp:946]: Codecov: Added lines #L944 - L946 were not covered by tests

File: src/librepgp/stream-armor.cpp [hunk 2]
```diff
@@ -956,20 +966,17 @@ armored_dst_write(pgp_dest_t *dst, const void *buf, size_t len)
         encptr += 4;
         param->lout += 4;
         if (param->lout == param->llen) {
-            if (param->usecrlf) {
-                *encptr++ = CH_CR;
-            }
-            *encptr++ = CH_LF;
+            armor_append_eol(param, encptr);
```
Comment (ronaldtse) [inline src/librepgp/stream-armor.cpp:969]: Codecov: Added line #L969 was not covered by tests

File: src/librepgp/stream-armor.h [hunk 3]
```diff
@@ -144,6 +145,30 @@ class ArmoredSource : public Source {
...
+
+  public:
+    ArmoredDest(const ArmoredDest &) = delete;
+    ArmoredDest(ArmoredDest &&) = delete;
+
+    ArmoredDest(pgp_dest_t &writedst, pgp_armored_msg_t msgtype) : Dest(), writedst_(writedst)
+    {
+        auto ret = init_armored_dst(&dst_, &writedst_, msgtype);
+        if (ret) {
+            throw rnp::rnp_exception(ret);
```
Comment (ronaldtse) [inline src/librepgp/stream-armor.h:160]: Codecov: Added line #L160 was not covered by tests

File: src/librepgp/stream-write.cpp [hunk 4]
```diff
@@ -1213,22 +1211,21 @@ cleartext_dst_finish(pgp_dest_t *dst)
...
-    for (auto &sinfo : param->siginfos) {
-        if ((ret = signed_write_signature(param, &sinfo, &armordst))) {
-            break;
+    try {
+        rnp::ArmoredDest armor(*param->writedst, PGP_ARMORED_SIGNATURE);
+        armor.set_discard(true);
+        for (auto &sinfo : param->siginfos) {
+            auto ret = signed_write_signature(param, &sinfo, &armor.dst());
+            if (ret) {
+                return ret;
```
Comment (ronaldtse) [inline src/librepgp/stream-write.cpp:1220]: Codecov: Added line #L1220 was not covered by tests

File: src/librepgp/stream-write.cpp [hunk 5]
```diff
@@ -1213,22 +1211,21 @@ cleartext_dst_finish(pgp_dest_t *dst)
...
+            }
         }
+        armor.set_discard(false);
+        return RNP_SUCCESS;
+    } catch (const std::exception &e) {
+        RNP_LOG("Failed to write armored signature: %s", e.what());
+        return RNP_ERROR_WRITE;
```
Comment (ronaldtse) [inline src/librepgp/stream-write.cpp:1227]: Codecov: Added lines #L1225 - L1227 were not covered by tests

**Review by ronaldtse — APPROVED** [review]
There are several codecov warnings. Feel free to merge if they are minor enough.

**Comment by ni4** [comment]
@ronaldtse thanks! Actually those codecov warning are for the lines which were not covered before, and are left for the case something really unusual happens, don't have an idea how to simulate those exception and got that covered.

**Review by andrey-utkin — APPROVED** [review]
LGTM.

Speculatively, I guess for write_transferable_keys() there could be an alternative resolution: do not expell it from the codebase, but promote it into public API. That function is not called from within the codebase, but it is called from the tests, and since the tests represent what people are doing/achieving in the real world, there could be some value in that function being shipped?

**Comment by ronaldtse** [comment]
Is this ready to be merged? Thanks!

**Comment by ni4** [comment]
As for me it seems ready (two approvals + CI passing), let's wait for @andrey-utkin reply.

**Comment by andrey-utkin** [comment]
> As for me it seems ready (two approvals + CI passing), let's wait for @andrey-utkin reply.

I'm happy for this to be merged.

**Pull request merged**